include/rocksdb/db.h

```diff
@@ -549,53 +549,62 @@ class DB {
   // any, or an empty value otherwise.
   //
   // If timestamp is enabled and a non-null timestamp pointer is passed in,
-  // timestamp is returned.
+  // timestamp is returned. If the underlying DB implementation doesn't
+  // support returning timestamp and the timestamp argument is non-null,
+  // an error will be returned.
   //
   // Returns OK on success. Returns NotFound and an empty value in "*value" if
   // there is no entry for "key". Returns some other non-OK status on error.
+  virtual Status Get(const ReadOptions& /*options*/,
+                     ColumnFamilyHandle* /*column_family*/,
+                     const Slice& /*key*/, PinnableSlice* /*value*/,
+                     std::string* /*timestamp*/) = 0;
+
+  // Similar to above, but returns the value in a string rather than
+  // PinnableSlice
   virtual inline Status Get(const ReadOptions& options,
                             ColumnFamilyHandle* column_family, const Slice& key,
-                            std::string* value) {
+                            std::string* value, std::string* timestamp) {
     assert(value != nullptr);
     PinnableSlice pinnable_val(value);
     assert(!pinnable_val.IsPinned());
-    auto s = Get(options, column_family, key, &pinnable_val);
+    auto s = Get(options, column_family, key, &pinnable_val, timestamp);
     if (s.ok() && pinnable_val.IsPinned()) {
       value->assign(pinnable_val.data(), pinnable_val.size());
     }  // else value is already assigned
     return s;
   }
+
+  // DEPRECATED: This function might be removed in a future release
   virtual Status Get(const ReadOptions& options,
                      ColumnFamilyHandle* column_family, const Slice& key,
-                     PinnableSlice* value) = 0;
+                     PinnableSlice* value) final {
+    return Get(options, column_family, key, value, nullptr);
+  }
+
+  // DEPRECATED: This function might be removed in a future release
   virtual Status Get(const ReadOptions& options, const Slice& key,
-                     std::string* value) {
+                     std::string* value) final {
     return Get(options, DefaultColumnFamily(), key, value);
   }
```

Comment:
Thanks for working on cleaning up all these overrides! I think changing the fallback logic so we implement older/less general APIs in terms of newer/more general ones, and having the fallback logic itself in the base class/interface (as opposed to repeating it in derived classes) makes perfect sense. There are few things I'm wondering about though:
1) Not sure if we would actually want to deprecate/remove the timestamp-less read APIs. From a theoretical perspective, user-defined timestamps are an optional part of our data model, and from a practical standpoint, even if we wanted to remove these APIs, migrating all existing users might not be realistic.
2) I think we might have some lower hanging fruit where a newer API provides a clearly better experience across the board: e.g. we might consider deprecating/removing the old-style non-batched `MultiGet` altogether, and potentially the point lookup variants that don't take a `ColumnFamilyHandle` and the ones that take `string` output parameters (although the latter two might be also in widespread usage).
3) On the technical side, our fallback methods in general could be regular non-virtual functions (instead `virtual` + `final`). This would also enable keeping the interface while eliminating the possibility of derived classes overriding some but not all variants.

Reply:
I'm also reluctant to deprecate functions that well represent a common subset of the data model. There's nothing wrong with convenience wrappers, especially if we are sure enough about their meaning in terms of other functions that we eliminate the option to override them.

Re: 3: See https://stackoverflow.com/a/11704714
> One reason some give for a base class to declare a non-overriding method to be final is simply so that anyone trying to define that method in a derived class gets an error instead of silently creating a method that 'hides' the base class's method.

Given that it was previously overridable, `virtual final` seems like the clear choice even though its weird at first glance.

Reply:
> Not sure if we would actually want to deprecate/remove the timestamp-less read APIs. From a theoretical perspective, user-defined timestamps are an optional part of our data model, and from a practical standpoint, even if we wanted to remove these APIs, migrating all existing users might not be realistic.

> I'm also reluctant to deprecate functions that well represent a common subset of the data model. There's nothing wrong with convenience wrappers, especially if we are sure enough about their meaning in terms of other functions that we eliminate the option to override them.

Makes sense. I went back and forth on these, but I agree there's some utility to having the timestamp-less variants in the user facing API. I'll not mark them as deprecated, but leave them as `final` so they cannot be overridden.

Reply:
Regarding declaring a function as `virtual final` vs non-virtual, is there a generally prescribed way of doing it? I guess its six of one half a dozen of the other. Personally, I like `virtual final` as it makes it pretty obvious it should not be overridden.